File: client-apps/cli/internal/cli/types/aliases_test.go
package types

import (
	"testing"
)

func TestGenerateAliases_McpServer(t *testing.T) {
	aliases := GenerateAliases("McpServer", "MCP Server", "mcp")

	// These are the unique lowercase forms that should be generated
	// The lookup is case-insensitive, so we check normalized forms
	expected := map[string]bool{
		"mcpserver":   true, // lowercase(name)
		"mcp-server":  true, // kebab-case(name)
		"mcp_server":  true, // snake_case(name)
		"mcp":         true, // display_name word[0] / idPrefix
		"mcpservers":  true, // pluralize(mcpserver)
		"mcp-servers": true, // pluralize(mcp-server)
		"mcp_servers": true, // pluralize(mcp_server)
		"mcps":        true, // pluralize(mcp)
	}

	found := make(map[string]bool)
	for _, alias := range aliases {
		found[NormalizeAlias(alias)] = true
	}

	for exp := range expected {
		if !found[exp] {
			t.Errorf("expected alias %q not generated", exp)
		}
	}
}

func TestGenerateAliases_Agent(t *testing.T) {
	aliases := GenerateAliases("Agent", "Agent", "agt")

	// Check normalized (lowercase) forms
	expected := map[string]bool{
		"agent":  true,
		"agt":    true,
		"agents": true,
		"agts":   true,
	}

	found := make(map[string]bool)
	for _, alias := range aliases {
		found[NormalizeAlias(alias)] = true
	}

	for exp := range expected {
		if !found[exp] {
			t.Errorf("expected alias %q not generated", exp)
		}
	}
}

func TestGenerateAliases_Workflow(t *testing.T) {
	aliases := GenerateAliases("Workflow", "Workflow", "wfl")

	// Check normalized (lowercase) forms
	expected := map[string]bool{
		"workflow":  true,
		"wfl":       true,
		"workflows": true,
		"wfls":      true,
	}

	found := make(map[string]bool)
	for _, alias := range aliases {
		found[NormalizeAlias(alias)] = true
	}

	for exp := range expected {
		if !found[exp] {
			t.Errorf("expected alias %q not generated", exp)
		}
	}
}

func TestGenerateAliases_NoDuplicates(t *testing.T) {
	aliases := GenerateAliases("Agent", "Agent", "agent")

	seen := make(map[string]int)
	for _, alias := range aliases {
		lower := NormalizeAlias(alias)
		seen[lower]++
	}

	for alias, count := range seen {
		if count > 1 {
			t.Errorf("duplicate alias %q found %d times", alias, count)
		}
	}
}

func TestToKebabCase(t *testing.T) {
	tests := []struct {
		input    string
		expected string
	}{
		{"McpServer", "mcp-server"},
		{"Agent", "agent"},
		{"AgentInstance", "agent-instance"},
		{"WorkflowExecution", "workflow-execution"},
		{"", ""},
	}

	for _, tt := range tests {
		t.Run(tt.input, func(t *testing.T) {
			result := toKebabCase(tt.input)
			if result != tt.expected {
				t.Errorf("toKebabCase(%q) = %q, expected %q", tt.input, result, tt.expected)
			}
		})
	}
}

func TestToSnakeCase(t *testing.T) {
	tests := []struct {
		input    string
		expected string
	}{
		{"McpServer", "mcp_server"},
		{"Agent", "agent"},
		{"AgentInstance", "agent_instance"},
		{"WorkflowExecution", "workflow_execution"},
		{"", ""},
	}

	for _, tt := range tests {
		t.Run(tt.input, func(t *testing.T) {
			result := toSnakeCase(tt.input)
			if result != tt.expected {
				t.Errorf("toSnakeCase(%q) = %q, expected %q", tt.input, result, tt.expected)
			}
		})
	}
}

func TestPluralize(t *testing.T) {
	tests := []struct {
		input    string
		expected string
	}{
		{"agent", "agents"},
		{"workflow", "workflows"},
		{"mcpserver", "mcpservers"},
		{"agents", "agents"}, // Already plural
		{"", ""},
	}

	for _, tt := range tests {
		t.Run(tt.input, func(t *testing.T) {
			result := pluralize(tt.input)
			if result != tt.expected {
				t.Errorf("pluralize(%q) = %q, expected %q", tt.input, result, tt.expected)
			}
		})
	}
}

func TestNormalizeAlias(t *testing.T) {
	tests := []struct {
		input    string
		expected string
	}{
		{"Agent", "agent"},
		{"  agent  ", "agent"},
		{"MCP-Server", "mcp-server"},
		{"WORKFLOW", "workflow"},
		{"", ""},
	}

	for _, tt := range tests {
		t.Run(tt.input, func(t *testing.T) {
			result := NormalizeAlias(tt.input)
			if result != tt.expected {
				t.Errorf("NormalizeAlias(%q) = %q, expected %q", tt.input, result, tt.expected)
			}
		})
	}
}

File: client-apps/cli/src/registry/aliases.test.ts
import { describe, expect, it } from "vitest";
import { generateAliases, normalizeAlias, pluralize, toKebabCase, toSnakeCase } from "./aliases.js";

function normalized(aliases: string[]): Set<string> {
  return new Set(aliases.map(normalizeAlias));
}

describe("generateAliases", () => {
  it("derives the full McpServer alias set", () => {
    const found = normalized(generateAliases("McpServer", "MCP Server", "mcp", "mcp_server"));
    for (const exp of [
      "mcpserver",
      "mcp-server",
      "mcp_server",
      "mcp",
      "mcpservers",
      "mcp-servers",
      "mcp_servers",
      "mcps",
    ]) {
      expect(found).toContain(exp);
    }
  });

  it("derives the Agent alias set", () => {
    const found = normalized(generateAliases("Agent", "Agent", "agt", "agent"));
    for (const exp of ["agent", "agt", "agents", "agts"]) {
      expect(found).toContain(exp);
    }
  });

  it("derives the Workflow alias set", () => {
    const found = normalized(generateAliases("Workflow", "Workflow", "wfl", "workflow"));
    for (const exp of ["workflow", "wfl", "workflows", "wfls"]) {
      expect(found).toContain(exp);
    }
  });

  it("derives the OAuthApp alias set including the canonical proto name", () => {
    // "OAuthApp" is the only kind whose PascalCase name has consecutive
    // capitals, so its split-derived forms (o-auth-app) diverge from the proto
    // enum name (oauth_app). Both families must be accepted: the canonical
    // spellings via protoName (stigmer/stigmer#470), and the historical
    // split-derived spellings for backward compatibility.
    const found = normalized(generateAliases("OAuthApp", "OAuth App", "oapp", "oauth_app"));
    for (const exp of [
      "oauth_app",
      "oauth-app",
      "oauth_apps",
      "oauth-apps",
      "oauthapp",
      "oauthapps",
      "o_auth_app",
      "o-auth-app",
      "oapp",
      "oapps",
    ]) {
      expect(found).toContain(exp);
    }
  });

  it("does not let a multi-word display name steal the parent's name", () => {
    // "Agent Instance" must NOT register "agent" (that belongs to Agent).
    const found = normalized(generateAliases("AgentInstance", "Agent Instance", "ain", "agent_instance"));
    expect(found).not.toContain("agent");
    expect(found).toContain("agentinstance");
    expect(found).toContain("agent-instance");
    expect(found).toContain("ain");
  });

  it("produces no duplicate normalized aliases", () => {
    // "agent" as protoName re-derives already-added forms — dedupe must hold.
    const aliases = generateAliases("Agent", "Agent", "agent", "agent");
    const lower = aliases.map(normalizeAlias);
    expect(new Set(lower).size).toBe(lower.length);
  });
});

describe("case conversion", () => {
  it.each([
    ["McpServer", "mcp-server"],
    ["Agent", "agent"],
    ["AgentInstance", "agent-instance"],
    ["WorkflowExecution", "workflow-execution"],
    ["", ""],
  ])("toKebabCase(%s) = %s", (input, expected) => {
    expect(toKebabCase(input)).toBe(expected);
  });

  it.each([
    ["McpServer", "mcp_server"],
    ["Agent", "agent"],
    ["AgentInstance", "agent_instance"],
    ["WorkflowExecution", "workflow_execution"],
    ["", ""],
  ])("toSnakeCase(%s) = %s", (input, expected) => {
    expect(toSnakeCase(input)).toBe(expected);
  });
});

describe("pluralize", () => {
  it.each([
    ["agent", "agents"],
    ["workflow", "workflows"],
    ["mcpserver", "mcpservers"],
    ["agents", "agents"],
    ["", ""],
  ])("pluralize(%s) = %s", (input, expected) => {
    expect(pluralize(input)).toBe(expected);
  });
});

describe("normalizeAlias", () => {
  it.each([
    ["Agent", "agent"],
    ["  agent  ", "agent"],
    ["MCP-Server", "mcp-server"],
    ["WORKFLOW", "workflow"],
    ["", ""],
  ])("normalizeAlias(%s) = %s", (input, expected) => {
    expect(normalizeAlias(input)).toBe(expected);
  });
});
